File: bro/src/main/java/me/xx2bab/bro/core/util/BroRudder.java
package me.xx2bab.bro.core.util;

import java.util.HashMap;
import java.util.Map;

import me.xx2bab.bro.common.Constants;
import me.xx2bab.bro.common.gen.GenOutputs;

public class BroRudder {

    private GenOutputs genOutputs;
    private Map<Class, Object> implCache;

    public BroRudder() {
        genOutputs = new GenOutputs();
        implCache = new HashMap<>();
    }

    @SuppressWarnings("unchecked")
    public <T> T getImplementationByInterface(Class<T> interfaze) {
        Object cacheRes = implCache.get(interfaze);
        if (cacheRes != null) {
            return (T) cacheRes;
        }
        try {
            String implClassName = genOutputs.generateClassNameForImplementation(interfaze);
            Object res = Class.forName(Constants.GEN_PACKAGE_NAME + "." + implClassName).newInstance();
            implCache.put(interfaze, res);
            return (T) res;
        } catch (Exception e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: bro/src/main/java/me/xx2bab/bro/core/util/BroRudder.kt
package me.xx2bab.bro.core.util

import me.xx2bab.bro.common.Constants
import me.xx2bab.bro.common.gen.GenOutputs.Companion.generateClassNameForImplementation
import java.util.*

class BroRudder {
    private val implCache: MutableMap<Class<*>, Any>
    @Suppress("UNCHECKED_CAST")
    fun <T> getImplementationByInterface(interfaze: Class<T>): T {
        val cacheRes = implCache[interfaze]
        return if (cacheRes != null) {
            cacheRes as T
        } else try {
            val implClassName = generateClassNameForImplementation(interfaze)
            val res = Class.forName(Constants.GEN_PACKAGE_NAME + "." + implClassName).newInstance()
            implCache[interfaze] = res
            res as T
        } catch (e: Exception) {
            throw IllegalArgumentException(e)
        }
    }

    init {
        implCache = HashMap()
    }
}